File: backend/app/api/lessons.py
"""Lesson endpoints: fetch a lesson, mark complete, update progress/quiz score.

The mobile app navigates to /course/:courseId/lesson/:lessonId and calls:
  GET    /lessons/{id}             — lesson content
  POST   /lessons/{id}/complete    — mark as complete
  PATCH  /lessons/{id}/progress    — update progress + optional quiz_score

Progress is persisted per (user, lesson) in the `lesson_progress` table
and the parent course's `enrollment.progress` is recomputed as the average
across all of the course's lessons, so the web dashboard and mobile client
both show real completion statistics.

All mutating endpoints require a valid access token (Bearer or cookie).
"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Optional
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel  # inline request body for PATCH progress
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.security import get_current_user_id  # auth guard on all mutations
from app.models import Enrollment, Lesson, LessonProgress, Module
from app.services import streak_service  # completing lessons feeds the streak

router = APIRouter()


class ProgressUpdate(BaseModel):
    """PATCH body: `progress` (0–100) and an optional quiz score."""
    progress: float
    quiz_score: Optional[float] = None


def _lesson_dict(lesson: Lesson, extra: Optional[dict] = None) -> dict:
    """Serialize a lesson to the JSON shape clients expect, optionally
    merged with extra fields (e.g. per-user progress)."""
    payload = {
        "id": str(lesson.id),
        "title": lesson.title,
        "content": lesson.content,
        "video_url": lesson.video_url,
        "duration": lesson.duration,
        "order": lesson.order,
        "is_published": lesson.is_published,
        "created_at": lesson.created_at.isoformat() if lesson.created_at else None,
    }
    if extra:
        payload.update(extra)
    return payload


async def _course_id_for_lesson(db: AsyncSession, lesson: Lesson) -> UUID | None:
    """Resolve the parent course of a lesson by hopping through its module."""
    module = await db.get(Module, lesson.module_id)
    return module.course_id if module else None


async def _upsert_lesson_progress(
    db: AsyncSession,
    user_id: UUID,
    lesson_id: UUID,
    progress: float,
    quiz_score: Optional[float] = None,
) -> LessonProgress:
    """Insert-or-update a single (user, lesson) progress row.

    A row is created when none exists yet; an existing row has its progress,
    completion flag, quiz score, and completed_at updated instead. The row
    is flushed (not committed) so the caller can commit once at the end.
    """
    result = await db.execute(
        select(LessonProgress).where(
            LessonProgress.user_id == user_id,
            LessonProgress.lesson_id == lesson_id,
        )
    )
    row = result.scalar_one_or_none()
    if row is None:
        row = LessonProgress(user_id=user_id, lesson_id=lesson_id, progress=0.0)
        db.add(row)
    # 100% progress flips the completion flag (and the timestamp).
    row.progress = progress
    row.is_completed = progress >= 100.0
    row.quiz_score = quiz_score
    row.completed_at = datetime.utcnow() if row.is_completed else None
    await db.flush()
    return row


async def _recompute_course_progress(
    db: AsyncSession,
    user_id: UUID,
    course_id: UUID,
) -> dict:
    """Average the user's lesson progress across a course and persist it on
    the enrollment row. Returns {percent, completed, total}."""
    # Every lesson that belongs to this course.
    lessons_result = await db.execute(
        select(Lesson).join(Module).where(Module.course_id == course_id)
    )
    lessons = list(lessons_result.scalars().all())
    total = len(lessons)
    if total == 0:
        return {"percent": 0.0, "completed": 0, "total": 0}

    # The user's progress rows for those lessons, keyed by lesson id.
    rows_result = await db.execute(
        select(LessonProgress).where(
            LessonProgress.user_id == user_id,
            LessonProgress.lesson_id.in_([lesson.id for lesson in lessons]),
        )
    )
    progress_rows = {row.lesson_id: row for row in rows_result.scalars().all()}
    completed = sum(
        1
        for lesson in lessons
        if progress_rows.get(lesson.id) and progress_rows[lesson.id].is_completed
    )
    # Course progress = mean of the per-lesson percentages.
    average = (
        sum(
            progress_rows.get(lesson.id).progress if progress_rows.get(lesson.id) else 0.0
            for lesson in lessons
        )
        / total
    )

    enrollment_result = await db.execute(
        select(Enrollment).where(
            Enrollment.user_id == user_id,
            Enrollment.course_id == course_id,
        )
    )
    enrollment = enrollment_result.scalar_one_or_none()
    if enrollment is not None:
        # Persist the recomputed average; stamp completion at 100%.
        enrollment.progress = round(average, 2)
        enrollment.completed_at = datetime.utcnow() if average >= 100.0 else None

    await db.commit()
    return {"percent": average, "completed": completed, "total": total}


@router.get("/{lesson_id}", response_model=dict)
async def get_lesson(lesson_id: uuid.UUID, db: AsyncSession = Depends(get_db)):
    """Return a single lesson with its content, duration, and order.
    Public read — no auth token required."""
    result = await db.execute(select(Lesson).where(Lesson.id == lesson_id))
    lesson = result.scalar_one_or_none()
    if lesson is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Lesson not found",
        )
    return _lesson_dict(lesson)


@router.post("/{lesson_id}/complete")
async def complete_lesson(
    lesson_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    user_id: uuid.UUID = Depends(get_current_user_id),
):
    """Mark a lesson as complete (progress 100%) and persist it.

    Recomputes the parent course's enrollment progress and records a
    learning day for the streak. Returns the lesson plus course progress.
    """
    result = await db.execute(select(Lesson).where(Lesson.id == lesson_id))
    lesson = result.scalar_one_or_none()
    if lesson is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Lesson not found",
        )

    course_id = await _course_id_for_lesson(db, lesson)
    # Always 100% here — that's what "complete" means.
    await _upsert_lesson_progress(db, user_id, lesson_id, 100.0)
    course_progress = (
        await _recompute_course_progress(db, user_id, course_id) if course_id else None
    )
    # Completing a lesson counts as a learning day for the streak.
    await streak_service.record_learning_day(db, user_id)

    return _lesson_dict(
        lesson,
        {
            "is_completed": True,
            "progress": 100.0,
            "course_progress_percent": int(course_progress["percent"]) if course_progress else None,
            "completed_lessons": course_progress["completed"] if course_progress else None,
            "total_lessons": course_progress["total"] if course_progress else None,
        },
    )


@router.patch("/{lesson_id}/progress")
async def update_lesson_progress(
    lesson_id: uuid.UUID,
    payload: ProgressUpdate,
    db: AsyncSession = Depends(get_db),
    user_id: uuid.UUID = Depends(get_current_user_id),
):
    """Persist lesson progress and optionally record a quiz score.

    Requires authentication. Recomputes the parent course's enrollment
    progress and returns the updated lesson with progress and quiz_score.
    """
    result = await db.execute(select(Lesson).where(Lesson.id == lesson_id))
    lesson = result.scalar_one_or_none()
    if lesson is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Lesson not found",
        )

    await _upsert_lesson_progress(
        db,
        user_id,
        lesson_id,
        payload.progress,
        quiz_score=payload.quiz_score,
    )
    course_id = await _course_id_for_lesson(db, lesson)
    course_progress = (
        await _recompute_course_progress(db, user_id, course_id) if course_id else None
    )
    # Meaningful progress (≥50%) also counts as a learning day.
    if payload.progress >= 50.0:
        await streak_service.record_learning_day(db, user_id)

    return _lesson_dict(
        lesson,
        {
            "is_completed": payload.progress >= 100.0,
            "progress": payload.progress,
            "quiz_score": payload.quiz_score,
            "course_progress_percent": int(course_progress["percent"]) if course_progress else None,
            "completed_lessons": course_progress["completed"] if course_progress else None,
            "total_lessons": course_progress["total"] if course_progress else None,
        },
    )

File: backend-node/src/routes/lessons.ts
/**
 * Lesson endpoints — mirrors backend/app/api/lessons.py.
 *
 * GET /lessons/:id (public), POST /lessons/:id/complete and
 * PATCH /lessons/:id/progress (authenticated). Progress is persisted per
 * (user, lesson) in `lesson_progress`; the parent course's `enrollments`
 * progress is recomputed as the average across its lessons, and completing a
 * lesson (or ≥50% progress) records a learning day for the streak.
 */
import { Router } from "express";
import { z } from "zod";

import { db } from "../db.js";
import type { Row } from "../db.js";
import { notFound, wrap } from "../errors.js";
import { requireUser, type AuthedRequest } from "../middleware/auth.js";
import { normalizeDt, nowIso } from "../serialize.js";
import { recordLearningDay } from "../services/streakService.js";
import { validate } from "../validate.js";

export const router = Router();

const ProgressSchema = z.object({
  progress: z.number(),
  quiz_score: z.number().nullish(),
});

interface LessonRow extends Row {
  id: string;
  module_id: string;
  title: string;
  content: string | null;
  video_url: string | null;
  duration: string | null;
  order: number;
  is_published: boolean;
  resource_links: unknown | null;
  created_at: string | null;
}

function lessonDict(l: LessonRow, extra?: Record<string, unknown>): Record<string, unknown> {
  const payload: Record<string, unknown> = {
    id: l.id,
    title: l.title,
    content: l.content,
    video_url: l.video_url,
    duration: l.duration,
    order: l.order,
    is_published: l.is_published,
    resource_links: l.resource_links ?? [],
    created_at: normalizeDt(l.created_at),
  };
  if (extra) Object.assign(payload, extra);
  return payload;
}

async function courseIdForLesson(lesson: LessonRow): Promise<string | null> {
  const module = await db.get<{ course_id: string }>(`SELECT course_id FROM modules WHERE id = $1`, [
    lesson.module_id,
  ]);
  return module ? module.course_id : null;
}

async function upsertLessonProgress(
  userId: string,
  lessonId: string,
  progress: number,
  quizScore?: number | null
): Promise<void> {
  const existing = await db.get<Row>(
    `SELECT id FROM lesson_progress WHERE user_id = $1 AND lesson_id = $2`,
    [userId, lessonId]
  );
  const isCompleted = progress >= 100.0;
  const completedAt = isCompleted ? nowIso() : null;
  if (existing) {
    await db.query(
      `UPDATE lesson_progress
          SET progress = $3, is_completed = $4, quiz_score = $5, completed_at = $6, updated_at = $7
        WHERE id = $1`,
      [existing.id, null, progress, isCompleted, quizScore ?? null, completedAt, nowIso()]
    );
  } else {
    await db.query(
      `INSERT INTO lesson_progress
         (id, user_id, lesson_id, progress, is_completed, quiz_score, completed_at, updated_at)
       VALUES ($1,$2,$3,$4,$5,$6,$7,$8)`,
      [crypto.randomUUID(), userId, lessonId, progress, isCompleted, quizScore ?? null, completedAt, nowIso()]
    );
  }
}

async function recomputeCourseProgress(
  userId: string,
  courseId: string
): Promise<{ percent: number; completed: number; total: number }> {
  const lessons = await db.query<{ id: string }>(
    `SELECT l.id FROM lessons l JOIN modules m ON l.module_id = m.id WHERE m.course_id = $1`,
    [courseId]
  );
  const total = lessons.length;
  if (total === 0) return { percent: 0.0, completed: 0, total: 0 };

  const progressRows = await db.query<Row>(
    `SELECT lesson_id, progress, is_completed FROM lesson_progress
      WHERE user_id = $1 AND lesson_id = ANY($2)`,
    [userId, lessons.map((l) => l.id)]
  );
  const byId = new Map(progressRows.map((r) => [r.lesson_id, r]));
  let completed = 0;
  let sum = 0;
  for (const l of lessons) {
    const row = byId.get(l.id);
    if (row && row.is_completed) completed += 1;
    sum += row ? Number(row.progress) : 0.0;
  }
  const average = sum / total;

  await db.query(
    `UPDATE enrollments
        SET progress = $3,
            completed_at = CASE WHEN $4::float >= 100 THEN $5 ELSE completed_at END
      WHERE user_id = $1 AND course_id = $2`,
    [userId, courseId, Math.round(average * 100) / 100, average, nowIso()]
  );
  return { percent: average, completed, total };
}

router.get(
  "/:lesson_id",
  wrap(async (req, res) => {
    const lesson = await db.get<LessonRow>(`SELECT * FROM lessons WHERE id = $1`, [
      req.params.lesson_id,
    ]);
    if (!lesson) throw notFound("Lesson not found");
    res.json(lessonDict(lesson));
  })
);

router.post(
  "/:lesson_id/complete",
  requireUser,
  wrap(async (req, res) => {
    const userId = (req as AuthedRequest).userId;
    const lesson = await db.get<LessonRow>(`SELECT * FROM lessons WHERE id = $1`, [
      req.params.lesson_id,
    ]);
    if (!lesson) throw notFound("Lesson not found");

    await upsertLessonProgress(userId, lesson.id, 100.0);
    const courseId = await courseIdForLesson(lesson);
    const courseProgress = courseId ? await recomputeCourseProgress(userId, courseId) : null;
    await recordLearningDay(userId);

    res.json(
      lessonDict(lesson, {
        is_completed: true,
        progress: 100.0,
        course_progress_percent: courseProgress ? Math.trunc(courseProgress.percent) : null,
        completed_lessons: courseProgress ? courseProgress.completed : null,
        total_lessons: courseProgress ? courseProgress.total : null,
      })
    );
  })
);

router.patch(
  "/:lesson_id/progress",
  requireUser,
  wrap(async (req, res) => {
    const userId = (req as AuthedRequest).userId;
    const data = validate(ProgressSchema, req.body);
    const lesson = await db.get<LessonRow>(`SELECT * FROM lessons WHERE id = $1`, [
      req.params.lesson_id,
    ]);
    if (!lesson) throw notFound("Lesson not found");

    // Clamp progress so a buggy/abusive client can't write values outside
    // 0–100 (which would corrupt the course progress average).
    const progress = Math.max(0, Math.min(Number(data.progress), 100));

    await upsertLessonProgress(userId, lesson.id, progress, data.quiz_score);
    const courseId = await courseIdForLesson(lesson);
    const courseProgress = courseId ? await recomputeCourseProgress(userId, courseId) : null;
    if (progress >= 50.0) await recordLearningDay(userId);

    res.json(
      lessonDict(lesson, {
        is_completed: progress >= 100.0,
        progress,
        quiz_score: data.quiz_score ?? null,
        course_progress_percent: courseProgress ? Math.trunc(courseProgress.percent) : null,
        completed_lessons: courseProgress ? courseProgress.completed : null,
        total_lessons: courseProgress ? courseProgress.total : null,
      })
    );
  })
);

/** ---------- Lesson quiz routes (end-of-lesson mastery self-check) ---------- */
import crypto from "crypto";

interface LessonQuiz {
  pass_percent: number;
  questions: { question: string; options: string[]; correct_index: number; explanation?: string }[];
}

const LessonResultSchema = z.object({
  score: z.number(),
  passed: z.boolean(),
  total_questions: z.number().int().default(0),
});

async function loadLessonForQuiz(lessonId: string): Promise<Row | null> {
  return db.get<Row>(`SELECT id, title, quiz FROM lessons WHERE id = $1`, [lessonId]);
}

function lessonQuizFor(lesson: Row): LessonQuiz | null {
  return (lesson.quiz as LessonQuiz | null) ?? null;
}

function lessonQuizJson(lesson: Row): Record<string, unknown> | null {
  const quiz = lessonQuizFor(lesson);
  if (!quiz) return null;
  return {
    lesson_id: lesson.id,
    title: lesson.title,
    pass_percent: quiz.pass_percent,
    questions: quiz.questions.map((q) => ({
      question: q.question,
      options: q.options,
      correct_index: q.correct_index,
      explanation: q.explanation ?? null,
    })),
  };
}

router.get(
  "/:lesson_id/quiz",
  requireUser,
  wrap(async (req, res) => {
    const lesson = await loadLessonForQuiz(req.params.lesson_id);
    if (!lesson) throw notFound("Lesson not found");
    const quiz = lessonQuizFor(lesson);
    if (!quiz) throw notFound("This lesson has no quiz");
    res.json(quiz);
  })
);

router.get(
  "/:lesson_id/quiz/result",
  requireUser,
  wrap(async (req, res) => {
    const userId = (req as AuthedRequest).userId;
    const lesson = await loadLessonForQuiz(req.params.lesson_id);
    if (!lesson) {
      res.json(null);
      return;
    }
    const quiz = lessonQuizFor(lesson);
    if (!quiz) {
      // No quiz authored for this lesson yet — treat as "no gate"
      res.json(null);
      return;
    }
    const row = await db.get<Row>(
      `SELECT id, score, passed, total_questions, created_at FROM quiz_results
        WHERE user_id = $1 AND lesson_id = $2
        ORDER BY created_at DESC LIMIT 1`,
      [userId, lesson.id]
    );
    if (!row) {
      res.json(null);
      return;
    }
    res.json({
      id: row.id,
      lesson_id: lesson.id,
      score: Number(row.score),
      passed: Boolean(row.passed),
      total_questions: Number(row.total_questions),
      created_at: normalizeDt(row.created_at as string | null),
    });
  })
);

router.post(
  "/:lesson_id/quiz/result",
  requireUser,
  wrap(async (req, res) => {
    const userId = (req as AuthedRequest).userId;
    const data = validate(LessonResultSchema, req.body);
    const lesson = await loadLessonForQuiz(req.params.lesson_id);
    if (!lesson) throw notFound("Lesson not found");
    const quiz = lessonQuizFor(lesson);
    if (!quiz) throw notFound("This lesson has no quiz");
    const total = data.total_questions || quiz.questions.length;
    const score = Math.max(0, Math.min(data.score, 100));
    const passed = score >= quiz.pass_percent;

    const rows = await db.query<{ created_at: string }>(
      `INSERT INTO quiz_results (id, user_id, lesson_id, score, passed, total_questions, created_at)
       VALUES ($1,$2,$3,$4,$5,$6,$7)
       RETURNING created_at`,
      [crypto.randomUUID(), userId, lesson.id, score, passed, total, new Date().toISOString()]
    );
    res.status(201).json({
      lesson_id: lesson.id,
      score,
      passed,
      total_questions: total,
      pass_percent: quiz.pass_percent,
      created_at: normalizeDt(rows[0]?.created_at ?? new Date().toISOString()),
    });
  })
);